objects/src/accounts/delta/vault.rs

```diff
@@ -37,6 +39,32 @@ impl AccountVaultDelta {
         }
     }
 
+    /// Merges another delta into this one, overwriting any existing values.
+    pub fn merge(self, other: Self) -> Self {
+        // Collec the assets into a hashmap where true indicates added, and false removed.
+        // This lets us simplify the addition/removal dance we would otherwise have to go through.
+        //
+        // Ordering matters. Since later items will overwrite earlier ones we need to begin
+        // with self, then other.
+        let assets = self
+            .added_assets
+            .into_iter()
+            .map(|asset| (asset, true))
+            .chain(self.removed_assets.into_iter().map(|asset| (asset, true)))
+            .chain(other.added_assets.into_iter().map(|asset| (asset, false)))
+            .chain(other.removed_assets.into_iter().map(|asset| (asset, false)))
+            .collect::<HashMap<_, _>>();
```

Comment:
I'm wondering why it isn't something like

```rust
pub struct AccountVaultDelta(HashMap<Asset, AssetDelta>);

enum AssetDelta {
  Added,
  Removed,
}
```

Reply:
I don't think there is a strong reason - probably was done like this to more closely follow the serialization format. Could you create an issue to consider refactoring this?

Reply:
https://github.com/0xPolygonMiden/miden-base/issues/798